### PR TITLE
Add note about filesystems to fs::rename

### LINKED ISSUE #24816: rename missing description details
[`std::fs::rename`](https://github.com/rust-lang/rust/blob/2fc2e12687ba7e48aef1c1d8efc6588a90e24010/src/libstd/fs.rs#L734) is missing the explicit description that you can also move a file outside of the current folder with it, as long as it's on the same file system.
This can be irritating.

### PR DESCRIPTION
**Pull request by steveklabnik**
Fixes #24816

r? @alexcrichton

### CODE REVIEW

File: src/libstd/fs.rs [hunk 1]
```diff
@@ -844,6 +844,8 @@ pub fn symlink_metadata<P: AsRef<Path>>(path: P) -> io::Result<Metadata> {
 
 /// Rename a file or directory to a new name.
 ///
+/// This will not work if the new name is on a different filesystem.
+///
```
Comment (Stebalien) [inline src/libstd/fs.rs]: Technically, on linux this will not work across mount points even if they are on the same filesystem.

Comment (tshepang) [inline src/libstd/fs.rs]: @Stebalien by "filesystem are the same", is ext4 an example of a filesystem?

Comment (Stebalien) [inline src/libstd/fs.rs]: Filesystem as used in this comment (i.e. partition?)

Comment (tshepang) [inline src/libstd/fs.rs]: In that you can have multiple mount points in the same partition?

Comment (Stebalien) [inline src/libstd/fs.rs]: I believe that depends on the filesystem. The trivial example is bind mount which lets you mount a directory:
```bash
mkdir /path/to/a
mkdir /path/to/b
mount -o bind /path/to/a /path/to/b
```
These both have the same filesystem (the directory `/path/to/a`) but different mount points. See the EXDEV comment in `man 2 rename` (on linux).

Comment (tshepang) [inline src/libstd/fs.rs]: Yeah, filesystem is rather ambiguous. Maybe @steveklabnik should simply say mount point?

Comment (Stebalien) [inline src/libstd/fs.rs]: In this case, filesystem is on linux (both interpretations) HOWEVER, on reading the windows spec, I get the impression that you can rename files between mount points (you can mount drives in folders: https://msdn.microsoft.com/en-us/library/windows/desktop/aa365733%28v=vs.85%29.aspx)  as long as they refer to the same underlying filesystem (partition).

Comment (tshepang) [inline src/libstd/fs.rs]: wow, impressive :)

**Pull request merged**